RecoBTag/PixelCluster/plugins/PixelClusterTagInfoProducer.cc

```diff
@@ -0,0 +1,162 @@
+#include "RecoBTag/PixelCluster/interface/PixelClusterTagInfoProducer.h"
+
+PixelClusterTagInfoProducer::PixelClusterTagInfoProducer(const edm::ParameterSet& iConfig)
+    : m_jets(consumes<edm::View<reco::Jet> >(iConfig.getParameter<edm::InputTag>("jets"))),
+      m_vertices(consumes<reco::VertexCollection>(iConfig.getParameter<edm::InputTag>("vertices"))),
+      m_pixelhit(consumes<edmNew::DetSetVector<SiPixelCluster> >(iConfig.getParameter<edm::InputTag>("pixelhit"))),
+      m_isPhase1(iConfig.getParameter<bool>("isPhase1")),
+      m_addFPIX(iConfig.getParameter<bool>("addForward")),
+      m_minADC(iConfig.getParameter<int>("minAdcCount")),
+      m_minJetPt(iConfig.getParameter<double>("minJetPtCut")),
+      m_maxJetEta(iConfig.getParameter<double>("maxJetEtaCut")) {
+  produces<reco::PixelClusterTagInfoCollection>();
+
+  m_nLayers = (m_isPhase1 ? 4 : 3);
+}
+
+PixelClusterTagInfoProducer::~PixelClusterTagInfoProducer() {}
+
+// ------------ method called to produce the data  ------------
+void PixelClusterTagInfoProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  // Declare produced collection
+  auto pixelTagInfo = std::make_unique<reco::PixelClusterTagInfoCollection>();
+
+  // Open jet collection
+  edm::Handle<edm::View<reco::Jet> > collectionJets;
+  iEvent.getByToken(m_jets, collectionJets);
+
+  // Count jets above threshold
+  int nJets(0);
+  for (auto jetIt = collectionJets->begin(); jetIt != collectionJets->end(); ++jetIt) {
+    if (jetIt->pt() > m_minJetPt)
+      nJets++;
+  }
+
+  // Get primary vertex in the event
+  edm::Handle<reco::VertexCollection> collectionPVs;
+  iEvent.getByToken(m_vertices, collectionPVs);
+  reco::VertexCollection::const_iterator firstPV = collectionPVs->begin();
+
+  for (reco::VertexCollection::const_iterator vtxIt = collectionPVs->begin(); vtxIt != collectionPVs->end(); ++vtxIt) {
+    firstPV = vtxIt;
+    break;
+  }
+
+  // If no suitable Jet and PV is available, skip the event without opening pixel collection
+  if (collectionPVs->empty() || nJets <= 0) {
+    iEvent.put(std::move(pixelTagInfo));
+    return;
+  }
+
+  // Open Pixel Cluster collection
+  edm::Handle<edmNew::DetSetVector<SiPixelCluster> > collectionClusters;
+  iEvent.getByToken(m_pixelhit, collectionClusters);
+
+  // Open Geometry
+  edm::ESHandle<TrackerGeometry> geom;
+  iSetup.get<TrackerDigiGeometryRecord>().get(geom);
+  const TrackerGeometry& theTracker(*geom);
+
+  // Retrieve tracker topology from geometry
+  edm::ESHandle<TrackerTopology> tTopoH;
+  iSetup.get<TrackerTopologyRcd>().get(tTopoH);
+  const TrackerTopology* tTopo = tTopoH.product();
+
+  std::vector<reco::PixelClusterProperties> clusters;
+
+  // Get vector of detunit ids, and fill a vector of PixelClusterProperties in the loop
+  for (edmNew::DetSetVector<SiPixelCluster>::const_iterator detUnit = collectionClusters->begin();
```

Comment:
It's certainly possible to use range loops, but since there are two nested loops we have to either:
1) make a copy of the DetSet<SiPixelCluster> object, slowing the code, or
2) have a not very readable code consisting of a chain of:
collectionClusters->data(d)->data(u) or even more complex syntax like
(*(*collectionClusters)[d])[u]

This is why we opted (and we still have the preference) for iterator loops in first place.

Reply:
```C++
  for (auto & detUnit :  collectionClusters) {
```
wouldn't work?

Reply:
Yes, but to make it work i have to first convert the edm::Handle<edmNew::DetSetVector<SiPixelCluster> >
to a
const edmNew::DetSetVector<SiPixelCluster>&